### PR TITLE
Addition of Hindi as Change Website Language.

### LINKED ISSUE #8734: Addition of Hindi as Change Website Language.
Addition of Hindi as Change Website Language. As there's a very vast no of people who know only Hindi so this integration  can help them move through the website easily without any language barrier.

THANKS.

[
Screenshot 2024-01-20 at 12 00 34 AM
](url)

### PR DESCRIPTION
**Pull request by rishabhkr-r111**
Closes #8734

Addition of Hindi as Change Website Language.
1. deleted ```openlibrary/i18n/hi/legacy-strings.hi.yml```
2. modifying ```openlibrary/templates/languages/language_list.html``` to add
   Hindi as a language.
3. adding and updating the ```hi/messages.po``` file to include Hindi translations.

note : As translations can be subjective, and my translation might not cover every possible thing ,but I feel it is a good start.

### Testing

Validated the translation
```docker compose exec -uroot web ./scripts/i18n-messages validate hi```
image

### Screenshot

image

### Stakeholders

@scottbarnes

### CODE REVIEW

**Comment by scottbarnes** [comment]
I will try to look more closely at this tomorrow, but I wanted to say thank you for the tremendous effort this must have took, @rishabhkr-r111.

File: openlibrary/i18n/hi/messages.po
Comment (scottbarnes) [inline openlibrary/i18n/hi/messages.po:4953]:
```suggestion
msgstr "एक और जोड़ें"
```

File: openlibrary/i18n/hi/messages.po
Comment (scottbarnes) [inline openlibrary/i18n/hi/messages.po:5048]:
```suggestion
msgstr "विवरण"
```

**Review by scottbarnes — CHANGES_REQUESTED** [review]
@rishabhkr-r111, being a mono-lingual English speaker, I did my best here, though I realize it's entirely possible my attempts to terminate sentences with `।` is completely wrong.

Insight on the plural-forms issue would be super helpful, though.

File: openlibrary/i18n/hi/messages.po
Comment (rishabhkr-r111) [reply openlibrary/i18n/hi/messages.po:20]: Traditional way to terminate the sentence is with ```।``` . However, nowadays, this has become more of a personal choose. Here are some snippets from some Indian national newspapers.
As you can in the first one the sentence terminate with ```.``` 
image
But, in the another newspaper's they will use ```।``` to terminate the sentences.
image
So, as long as we stick to one way (to maintain consistency), it should not be a problem.

### Some Insight on the plural-forms
After you mentioned this, I read more about plural-forms. 
```Plural-Forms: nplurals=2; plural=(n != 1);\n``` will be more correct choose.  In many cases, both expressions might yield similar results, but there can be some subtle differences in certain cases.
```
[0] = तुम्हारे पास 0 संदेश हैं।
[1] = तुम्हारे पास 1 संदेश है।
[2} = तुम्हारे पास 2 संदेश हैं।
```
You can see their is only difference of ```.``` in the [1] . 

@scottbarnes I hope you found this interesting. I tried my best to explain.

**Comment by rishabhkr-r111** [comment]
Done! I have made the necessary changes.

File: openlibrary/i18n/hi/messages.po
Comment (rishabhkr-r111) [reply openlibrary/i18n/hi/messages.po:103]: This is correct and will not require any changes . There are words in the Hindi language that remain the same in singular or plural, and sometimes it depends on the structure of sentences.
image

File: openlibrary/i18n/hi/messages.po
Comment (rishabhkr-r111) [reply openlibrary/i18n/hi/messages.po:121]: Here, it's better to change it.
```
msgid "Waiting for %d day"
msgid_plural "Waiting for %d days"
msgstr[0] "%d दिन की प्रतीक्षा"
msgstr[1] "%d दिनों की प्रतीक्षा"

```

**Comment by rishabhkr-r111** [comment]
> Is zero considered singular or plural in Hindi? In other words is तुम्हारे पास 0 संदेश हैं। correct, or should it be तुम्हारे पास 0 संदेश है।? The Unicode consortium seems to think singular https://www.unicode.org/cldr/charts/43/supplemental/language_plural_rules.html#hi which would imply something like `Plural-Forms: nplurals=2; plural=(n!=0 && n!=1);`. (Their examples are 1 घंटा and 2 घंटे which are more different). If zero is not singular then, then `plural=(n!=1)` would be more correct (ie the same as English).
> 
> Of course this is all just a convention, so you could use `plural=(n==0 || n==1)` and put the plural form first instead of second, but that seems confusing to me.
> 
> BTW, if the Unicode site is wrong, it would be good to submit a bug report to get it corrected so that it doesn't lead others astray (I don't speak Hindi, so I have no idea if it's right or not).

Zero considered singular in Hindi. However, usage of singular or plural forms can depend on the context and the nature of the noun being counted. Like,
image


The plural form might be used with certain nouns to emphasize the absence or lack of something.  Plural form "शून्य" (Zero) is less common than the singular form in regular speech. In most cases people will stick to the singular form for clarity and ease of pronunciation.

Usages of ```"Plural-Forms: nplurals=2; plural=(n != 1);\n"``` , I found was more common and accepted by most local  individual's when I searched (https://translations.launchpad.net/+languages/hi). But, we can go with  ```"Plural-Forms: nplurals=2; plural=(n==0 || n==1);\n"``` also which is used in ```poedit``` and that's was is mentioned by Unicode consortium (https://www.unicode.org/cldr/charts/43/supplemental/language_plural_rules.html#hi).

Both forms are in use, and opting for either should not pose a problem. Just we should document the choice to provide clarity to other contributors and maintainers.

**Comment by scottbarnes** [comment]
@rishabhkr-r111, @tfmorris, if I understand correctly, if we treat zero as singular, then, assuming `nplurals=2`:
- if we use `plural=(n==0 || n==1)`, we need to put the plural form first, but
- this format will be inconsistent with how these files are usually formatted; or
- we use `plural=(n!=0 && n!=1)` which allows us to preserve the 'usual' ordering of singular and plural, and accurately expresses plurals are numbers that are not `0` and not `1`.

One thing that has me puzzled is that, with respect to Hindi localizations, there appear to be around [13.2k files where plural=(n != 1)](https://github.com/search?q=%22Language%3A+hi%5Cn%22+%22%28n+%21%3D+1%29%22&type=code), and [107 where plural=(n==0 || n==1)](https://github.com/search?q=%22Language%3A+hi%5Cn%22+%22%28n%3D%3D0+%7C%7C+n%3D%3D1%29%22&type=code). And zero where `plural=(n!= 0 && n!= 1)`. There are also smaller numbers of files using different spacing around the various operators.

On the basis of sheer numbers alone, I'd be inclined to favor `"Plural-Forms: nplurals=2; plural=(n != 1);\n"`. Thoughts?

**Comment by tfmorris** [comment]
@rishabhkr-r111 What you wrote:

> Zero considered singular in Hindi. [...] Plural form "शून्य" (Zero) is less common than the singular form in regular speech. In most cases people will stick to the singular form for clarity and ease of pronunciation.

contradicts `"Plural-Forms: nplurals=2; plural=(n != 1);\n"` which will consider zero to be plural  (0 != 1 evaluates to `true` ie plural)

The poedit version `"Plural-Forms: nplurals=2; plural=(n==0 || n==1);\n"` is actually the opposite of the Unicode Consortium (as I interpret it), which would require `"Plural-Forms: nplurals=2; plural=!(n==0 || n==1);\n"` or equivalent

@scottbarnes Your comment arrived as I was writing. Your summary bullet points agree with my understanding.

I'm not sure popularity is the best metric, but there are another [936 instances](https://github.com/search?q=%22Language%3A+hi%5Cn%22+%22n+%3E+1%22&type=code) of `plural=n > 1` which is a simpler way to specify the Unicode Consortium criteria (if you ignore negative numbers). That's still only 1/13th of the popular case, but it matches @rishabhkr-r111 's description of zero being singular in addition to the Unicode Consortium's documentation. This is the form I'd favor.

**Comment by rishabhkr-r111** [comment]
@tfmorris  @scottbarnes I went through  ```Firefox``` official Hindi translation's. It seems they are also following ```"Plural-Forms: nplurals=2; plural=(n != 1);\n"``` .

https://pontoon.mozilla.org/hi-IN/firefox/all-resources/?status=translated&string=200662

image

Considering one(1) as - singular
Considering Others ( 0, 2 ) as - plural

```"Plural-Forms: nplurals=2; plural=(n != 1);\n"```   option is the most popular and is accepted by most locals, so I would like to  agree with @scottbarnes  and go with it

**Comment by tfmorris** [comment]
Sounds good. I'd suggest adding a comment explicitly referencing that the Unicode Consortium was considered and rejected so that people understand the rationale if it's questioned in the future.

**Comment by rishabhkr-r111** [comment]
> Sounds good. I'd suggest adding a comment explicitly referencing that the Unicode Consortium was considered and rejected so that people understand the rationale if it's questioned in the future.

```
"Content-Type: text/plain; charset=utf-8\n"
"Content-Transfer-Encoding: 8bit\n"
"Plural-Forms: nplurals=2; plural=(n != 1);\n"
# Note: The Unicode Consortium guidelines were considered, and this specific
# pluralization rule was chosen to align with common usage and linguistic
# characteristics in Hindi. The alternative rule "plural=(n==0 || n==1);" was
# rejected as the chosen form was found to be more widely accepted, providing
# clarity and consistency.
"Generated-By: Babel 2.9.1\n"
"X-Generator: Poedit 3.4.2\n"
```
@tfmorris Should I go with this comment? What is your opinion?

File: openlibrary/i18n/hi/messages.po
Comment (tfmorris) [reply openlibrary/i18n/hi/messages.po:121]: Can you update the translation then please?

**Comment by tfmorris** [comment]
I would omit the last sentence
~~The alternative rule "plural=(n==0 || n==1);" was~~
~~# rejected as the chosen form was found to be more widely accepted, providing~~
~~# clarity and consistency.~~

I would suggest updating the PR with your changes in one or more new commits (both this and the pluralization change that you suggested in an earlier comment) and then people can review them in context (and use Github's suggestion feature)

**Comment by rishabhkr-r111** [comment]
I've added the comment and made a few fixes. It should be good to go.

**Review by tfmorris — COMMENTED** [review]
Thanks for adding the comment! Hopefully that'll help future contributors.

I noticed that there are a number of translations missing. I started marking them, but then wasn't sure whether it was perhaps intentional. I think it's fine to merge this as a partial translation if that's the case. It represents a LOT of work (thank you!) and it would be good to make it available to people.

**Comment by rishabhkr-r111** [comment]
I believe merging this as a partial translation is a good idea. Making it available to users in its current state would definitely be beneficial and would also allow and promote others to contribute further.

**Comment by rishabhkr-r111** [comment]
@tfmorris finally got some time, and I updated few missing translations.

**Comment by tfmorris** [comment]
@rishabhkr-r111 That's great, but I don't have commit rights to this repo, so you'll need to wait for @scottbarnes are another committer to merge this.

**Review by tfmorris — APPROVED** [review]

**Review by scottbarnes — APPROVED** [review]
This looks good to me. Thanks @rishabhkr-r111 and @tfmorris. Sorry for the delay. I was pointlessly fretting over the pluralization, as I take the point that it's not a popularity contest.

**Pull request merged**
